external-crates/move/crates/move-vm-runtime/src/execution/interpreter/locals.rs

```diff
@@ -0,0 +1,305 @@
+// Heap
+// -------------------------------------------------------------------------------------------------
+
+/// The Move VM's base heap. This is PTBs and arguments to invocation functions are stored, so that
+/// we can handle references to/from them.
+#[derive(Debug)]
+pub struct BaseHeap {
+    next_id: usize,
+    values: HashMap<BaseHeapId, Box<Value>>,
+}
+
+/// An ID for an entry in a Base Heap.
+#[derive(Clone, Copy, Debug, PartialOrd, Ord, PartialEq, Eq, Hash)]
+pub struct BaseHeapId(usize);
+
+/// The runtime machine "heap" for execution. This allows us to grab and return frame slots and the
+/// like. Note that this isn't a _true_ heap (crrently), it only allows for allocating and freeing
+/// stackframes.
+#[derive(Debug)]
+pub struct MachineHeap {
+}
+
+/// A stack frame is an allocated frame. It was allocated starting at `start` in the heap. When it
+/// is freed, we need to check that we are freeing the one on the end of the heap.
+#[derive(Debug)]
+pub struct StackFrame {
+    slice: Vec<Value>,
+}
+
+// -------------------------------------------------------------------------------------------------
+// Base (Machine-External) Heap
+// -------------------------------------------------------------------------------------------------
+
+impl Default for BaseHeap {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+impl BaseHeap {
+    pub fn new() -> Self {
+        Self {
+            values: HashMap::new(),
+            next_id: 0,
+        }
+    }
+
+    /// Allocate a slot for the value in the base heap, and then
+    pub fn allocate_and_borrow_loc(
+        &mut self,
+        value: Value,
+    ) -> PartialVMResult<(BaseHeapId, Value)> {
+        let next_id = BaseHeapId(self.next_id);
+        self.next_id += 1;
+        self.values.insert(next_id, Box::new(value));
+        let ref_ = self.borrow_loc(next_id)?;
+        Ok((next_id, ref_))
+    }
+
+    /// Moves a location out of memory, swapping it with `ValueImpl::Invalid`
+    pub fn take_loc(&mut self, ndx: BaseHeapId) -> PartialVMResult<Value> {
+        if self.is_invalid(ndx)? {
+            return Err(
+                PartialVMError::new(StatusCode::UNKNOWN_INVARIANT_VIOLATION_ERROR)
+                    .with_message("Cannot move from an invalid memory location".to_string()),
+            );
+        }
+
+        let Some(value_box) = self.values.get_mut(&ndx) else {
+            return Err(
+                PartialVMError::new(StatusCode::UNKNOWN_INVARIANT_VIOLATION_ERROR)
+                    .with_message(format!("Invalid index: {}", ndx)),
+            );
+        };
+
+        let value = std::mem::replace(value_box.as_mut(), Value::invalid());
+        Ok(value)
+    }
+
+    /// Borrows the specified location
+    pub fn borrow_loc(&self, ndx: BaseHeapId) -> PartialVMResult<Value> {
+        self.values
+            .get(&ndx)
+            .ok_or_else(|| {
+                PartialVMError::new(StatusCode::INTERNAL_TYPE_ERROR)
```

Comment:
Note for the future: Feels odd that that is an `INTERNAL_TYPE_ERROR` while the others are `UNKNOWN_INVARIANT_VIOLATION`s. Maybe something to go through when doing an errors pass together (or I'm happy to take a first pass as well).